src/Vanilla/VanillaConnect/VanillaConnectProvider.php

```diff
@@ -0,0 +1,195 @@
+ */
+class VanillaConnectProvider {
+
+    /**
+     * Regex used to validate the whitelist URLs which can contain wildcards.
+     * It is a pretty loose regex that should enforce what is needed without blocked weird cases.
+     *
+     * One line regex:
+     * /^(?<scheme>(?:https?:)?\/\/)(?:[^\s]+?@)?(?<host>[^\/\s]+)(?<path>\/[^?#\s]*)$/
+     *
+     * @var string
+     */
+    protected static $schemeHostPathRegex =
+        '^' // Start of the line
+        .'(?<scheme>(?:https?:)?\/\/)' // Must start with http://, https:// or //
+        .'(?<userpwd>[^\s]+?@)?' // user:password@
+        .'(?<host>[^@\/\s]+)' // We allow pretty much everything until we encounter the path
+        .'(?<path>\/[^?#\s]*)' // The path. / is mandatory but the rest is optional.
+        .'$'
+    ;
+
+    /**
+     * @var array
+     */
+    private $allowedWhitelist;
+
+    /**
+     * @var VanillaConnect
+     */
+    private $vanillaConnect;
+
+    /**
+     * VanillaConnectProvider constructor.
+     *
+     * @param string S$clientID
+     * @param string $secret
+     * @param array $redirectURLsWhitelist
+     */
+    public function __construct($clientID, $secret, array $redirectURLsWhitelist) {
+        $this->vanillaConnect = new VanillaConnect($clientID, $secret);
+        $this->allowedWhitelist = $this->parseWhitelistURLs($redirectURLsWhitelist);
+    }
+
+    /**
+     * Create a response URL, from an authentication JWT and a claim.
+     *
+     * @param string $requestJWT JWT sent during the authentication request.
+     * @param array $claim The data to put as the claim in the the response JWT. Needs to contain id.
+     * @return string The URL to redirect to so that the response can be processed.
+     * @throws Exception
+     */
+    public function createResponseURL($requestJWT, array $claim) {
+        $errors = [];
+
+        $redirect = VanillaConnect::extractItemFromClaim($requestJWT, 'redirect');
+        if (empty($redirect)) {
+            throw new Exception('The authentication JWT claim is missing the "redirect" field.');
+        } else {
+            if (!$this->validateRedirectURL($redirect)) {
+                $errors['request_invalid_redirect'] = "The redirect URL '$redirect' is not whitelisted.";
+
+                $urlEncodingError = false;
+
+                // Common URL encoding error
+                if (strpos($redirect, ' ') !== false) {
+                    $urlEncodingError = true;
+                }
+
+                if ($urlEncodingError) {
+                    $errors['request_invalid_redirect_tip'] =
+                        "Seems like the redirect URL was not properly encoded. Invalid character detected.";
+                }
+            }
+        }
+
+        if ($this->vanillaConnect->validateRequest($requestJWT, $authClaim)) {
+            $nonce = $authClaim['nonce'];
+        } else {
+            $errors = $this->vanillaConnect->getErrors();
```

Comment:
We're recording specific errors in this function and then potentially overwriting all of them with whatever comes back from `VanillaConnect`.